EssentialApp/EssentialAppTests/ImageCommentsUIIntegrationTests.swift

```diff
@@ -0,0 +1,259 @@
+		
+		sut.simulateUserInitiatedReload()
+		loader.completeImageCommentLoading(at: 1)
+		XCTAssertEqual(sut.errorMessage, nil)
+	}
+	
+	// MARK: - Helpers
+	private func makeSUT() -> (sut: ImageCommentsViewController, loader: LoaderSpy) {
+		let loader = LoaderSpy()
+		let sut = ImageCommentsUIComposer.imageCommentsComposedWith(loader: loader)
+		trackForMemoryLeaks(loader)
+		trackForMemoryLeaks(sut)
+		return (sut, loader)
+	}
+	
+	private func makeImageComment(id: UUID, message: String, createdAt: (date: Date, formatted: String), username: String) -> (model: ImageComment, viewModel: ImageCommentViewModel) {
+		let model = ImageComment(
+			id: id,
+			message: message,
+			createdAt: createdAt.date,
+			author: ImageCommentAuthor(username: username))
+		let viewModel = ImageCommentViewModel(message: message, created: createdAt.formatted, username: username)
+		return (model, viewModel)
+	}
+	
+	func assertThat(_ sut: ImageCommentsViewController, isRendering imageComments: [ImageCommentViewModel], file: StaticString = #filePath, line: UInt = #line) {
+		sut.view.enforceLayoutCycle()
+		
+		guard sut.numberOfRenderedImageCommentViews() == imageComments.count else {
+			return XCTFail("Expected \(imageComments.count) image comments, got \(sut.numberOfRenderedImageCommentViews()) instead.", file: file, line: line)
+		}
+		
+		imageComments.enumerated().forEach { index, imageComment in
+			assertThat(sut, hasViewConfiguredFor: imageComment, at: index, file: file, line: line)
+		}
+		
+		executeRunLoopToCleanUpReferences()
+	}
+	
+	func assertThat(_ sut: ImageCommentsViewController, hasViewConfiguredFor imageComment: ImageCommentViewModel, at index: Int, file: StaticString = #filePath, line: UInt = #line) {
+		let view = sut.imageCommentView(at: index)
+		
+		guard let cell = view as? ImageCommentCell else {
+			return XCTFail("Expected \(ImageCommentCell.self) instance, got \(String(describing: view)) instead", file: file, line: line)
+		}
+		
+		XCTAssertEqual(cell.messageText, imageComment.message, "for view at index (\(index))", file: file, line: line)
+		XCTAssertEqual(cell.createdText, imageComment.created, "for view at index (\(index))", file: file, line: line)
+		XCTAssertEqual(cell.usernameText, imageComment.username, "for view at index (\(index))", file: file, line: line)
+	}
+	
+	private func executeRunLoopToCleanUpReferences() {
+		RunLoop.current.run(until: Date())
+	}
+	
+	func localized(_ key: String, file: StaticString = #filePath, line: UInt = #line) -> String {
+		let table = "ImageComments"
+		let bundle = Bundle(for: ImageCommentPresenter.self)
+		let value = bundle.localizedString(forKey: key, value: nil, table: table)
+		if value == key {
+			XCTFail("Missing localized string for key: \(key) in table: \(table)", file: file, line: line)
+		}
+		return value
+	}
+	
+	private class LoaderSpy: ImageCommentLoader {
+		
+		var completions = [(ImageCommentLoader.Result) -> Void]()
```

Comment:
This property's setter should be private to prevent clients from mutating it - it should only be mutated by the class instance.

```suggestion
		private (set) var completions = [(ImageCommentLoader.Result) -> Void]()
```